interfaces/builtin/unity8.go

```diff
@@ -0,0 +1,190 @@
+       member=DispatchURL
+       peer=(name=com.canonical.URLDispatcher,label=###SLOT_SECURITY_TAGS###),
+
+  # This is needed when the app is already running and needs to be passed in
+  # a URL to open. This is most often used with content-hub providers and
+  # url-dispatcher, but is actually supported by Qt generally.
+  dbus (receive)
+       bus=session
+       path=/###PLUG_DBUS_APPIDS###
+       interface=org.freedesktop.Application
+       member=Open
+       peer=(label=###SLOT_SECURITY_TAGS###),
+
+  # Unity launcher (e.g. app counter, progress, alert)
+  dbus (receive, send)
+       bus=session
+       path=/com/canonical/Unity/Launcher/###PLUG_DBUS_APPIDS###
+       peer=(name=com.canonical.Unity.Launcher,label=###SLOT_SECURITY_TAGS###),
+
+  # Content Hub (pasteboard, file transfers)
+  dbus (receive, send)
+       bus=session
+       interface=com.ubuntu.content.dbus.Service
+       path=/
+       peer=(name=com.ubuntu.content.dbus.Service,label=###SLOT_SECURITY_TAGS###),
+
+  # Lttng tracing is very noisy and should not be allowed by confined apps.
+  # Can safely deny. LP: #1260491
+  deny /{,var/}{dev,run}/shm/lttng-ust-* r,
+`)
+
+var unity8ConnectedPlugSecComp = []byte(`
+recvfrom
+recvmsg
+sendmsg
+sendto
+shutdown
+`)
+
+type Unity8Interface struct{}
+
+func (iface *Unity8Interface) Name() string {
+	return "unity8"
+}
+
+func (iface *Unity8Interface) String() string {
+	return iface.Name()
+}
+
+func (iface *Unity8Interface) dbusAppId(app *snap.AppInfo) []byte {
+	var retval bytes.Buffer
+
+	appidbits := []string{app.Snap.Name(), app.Name, app.Snap.Revision.String()}
+	appid := strings.Join(appidbits, "_")
+
+	for _, value := range appid {
+		if unicode.In(rune(value), unicode.Letter, unicode.Digit) {
+			retval.WriteRune(rune(value))
+		} else {
+			retval.WriteString(fmt.Sprintf("_%2.2x", value))
+		}
+	}
+
+	return retval.Bytes()
+}
+
+func (iface *Unity8Interface) PermanentPlugSnippet(plug *interfaces.Plug, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+func (iface *Unity8Interface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		oldTags := []byte("###SLOT_SECURITY_TAGS###")
+		newTags := slotAppLabelExpr(slot)
+		snippet := bytes.Replace(unity8ConnectedPlugAppArmor, oldTags, newTags, -1)
+
+		appidsOld := []byte("###PLUG_DBUS_APPIDS###")
+		var appidsNew bytes.Buffer
+		var appidsList []string
+		for _, app := range plug.Apps {
+			appidsList = append(appidsList, string(iface.dbusAppId(app)))
+		}
+		sort.Strings(appidsList) // makes tests reliable
+		if len(appidsList) == 1 {
+			appidsNew.WriteString(appidsList[0])
+		} else if len(appidsList) > 1 {
+			appidsNew.WriteByte('{')
+			appidsNew.WriteString(strings.Join(appidsList, ","))
+			appidsNew.WriteByte('}')
+		}
+		snippet = bytes.Replace(snippet, appidsOld, appidsNew.Bytes(), -1)
```

Comment:
+1 on adding a glob while the details are worked out.

I agree that snapd in general shouldn't care about what snaps are doing on top and I wouldn't have expressed on opinion on this at all except for the fact that Unity8's concept of AppID trickled into the snapd interface code, which alerted me to maintenance concerns, potential problems and developer confusion. Importantly, trusted helpers on Touch used the apparmor label as the AppID and could trust that. If you are using a different AppID for Personal, then the apparmor label won't correspond to that and the implementation will be considerably more complex.

While unity8 is in some ways 'just a snap', Ubuntu Personal should be working in concert with snapd, not against it. Areas where snapd currently is not meeting Personal's needs should be closely examined, perhaps with redesign. This will only help us solve problems for other desktop environments moving forward.

As for handling running applications while they're upgraded, remove or installed, that is a separate issue and UAL's current behavior is just one way of managing that, but it is incomplete (see https://bugs.launchpad.net/snappy/+bug/1616650 for one example). We should strive to solve those issues in snapd independently of UAL/unity8 interface so that all snaps will benefit.